Add Bard Song: Pining Nocturne

RaraProjects opened the pull request:
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?
Closes #4778

1. I added the Pining Nocturne (472) entry to spell_list.sql.
2. I implemented the spell Pining Nocturne in pining_nocturne.lua

The Pining Nocturne implementation was done generally as I hope to centralize the BRD song handling (both buffs and debuffs) in future work. My change is isolated to Pining Nocturne and no other spells should be affected at this time. I looked through the various other song implementations as well as the handling done in /scripts/globals/spells/enhancing_song.lua to come up with my strategy. Although Pining Nocturne can be applied successfully in game at this point. I don't think it will actually affect the mob as I wasn't able to find EFFECT_NOCTURNE (223), EFFECT_ADDLE (21), or EFFECT_MAGIC_ACC_DOWN (174) in the general spell cast handler. I looked at /src/map/ai/states/magic_state.cpp and /scripts/globals/magic.lua.

## Steps to test these changes
Cast Pining Nocturne on a mob with varying levels of Song+ gear, Soul Voice, Marcato, and Troubadour.

### Testing Results
Base Tests | Power (Wiki) | Duration (Wiki) | Result
-- | -- | -- | --
Base | 15.0 (15) | 120 (120) | Pass
All Songs +1   (Briosos Whistle) | 16.5 (16.5) | 132 (+10%) | Pass
All Songs +2   (Moonbow Whistle) | 18.0 (18) | 144 (+20%) | Pass
All Songs +3   (Moonbow Whistle +1) | 19.5 (19.5) | 156 (+30%) | Pass
All Songs +7   (Moonbow Whistle +1 + Gjallarhorn) | 25.5 (25.5) | 204 (+70%) | Pass

Soul Voice Tests (x2 Power) | Power (Wiki) | Duration (Wiki) | Result
-- | -- | -- | --
Base | 30.0 | 120 | Pass
All Songs +7   (Moonbow Whistle +1 + Gjallarhorn) | 51.0 | 204 | Pass

Marcato Tests (x1.5 Power) | Power (Wiki) | Duration (Wiki) | Result
-- | -- | -- | --
Base | 22.5 (+50%) | 120 | Pass
All Songs +7   (Moonbow Whistle +1 + Gjallarhorn) | 38.25 (+50%) | 204 | Pass

Troubadour Tests (x2 Duration) | Power (Wiki) | Duration (Wiki) | Result
-- | -- | -- | --
Base | 15.0 | 240 | Pass
All Songs +7   (Moonbow Whistle +1 + Gjallarhorn) | 25.5 | 408 | Pass

Soul Voice + Troubadour Tests | Power (Wiki) | Duration (Wiki) | Result
-- | -- | -- | --
Base | 30.0 | 240 | Pass
All Songs +7   (Moonbow Whistle +1 + Gjallarhorn) | 51.0 | 408 | Pass

## Additional Information
I confirmed that Pining Nocturne can be magic bursted in retail and have thus included the magic burst message in my SQL entry. I did not implement the magic burst message in my implementation.
![NocturneMagicBurst](https://github.com/user-attachments/assets/cb670e44-5941-49f0-a37f-9ed0b9b94223)

Here is the packet capture I used to determine animation. Animation = 295, Param = 223, Message = 268
![PacketCapture](https://github.com/user-attachments/assets/424ae978-90aa-4b7b-af12-bd08ef9d0d08)

Here is the SQL entry breakdown.
Column | Value | Comments
-- | -- | --
spellid | 472 |  
name | 'pining_nocturne' |  
jobs | 0x0000000000000000005F000000000000000000000000, | Piece 10 (BRD) is set to 95, the level of the spell.
group | 1 | BRD songs
family | 0 | Standalone   spell--not in a family.
element | @ELEMENT_FIRE | per Wiki
zonemisc | 0 | Matches other   songs.
validTargets | 4 | Matches other   debuff songs.
skill | @SKILL_SINGING | Matches other   songs.
mpCost | 0 | Matches other   songs.
castTime | 2000 | per Wiki (2   seconds)
recastTime | 24000 | per Wiki (24   seconds)
message | 237 | <actor>   casts <spell>. <target> receives the effect of <status>.
magicBurstMessage | 268 | Tested that this   can magic burst on retail.     Did not implement the message in the spell.
animation | 295 | Tested this on   retail.
animationTime | 2000 | Matches other   songs.
AOE | 0 | Matches other   debuff songs.
base | 0 | Matches other   songs.
multiplier | 1.0 | Matches other   songs.
CE | 20 | per Wiki
VE | 140 | per Wiki
requirements | 0 | Matches other   songs.
spell_range | 204 | Matches other   songs besides Requiem and Virelai.
content_tag | 'ABYSSEA' | Not sure, but   since the spell is 95 I believe that means it came out before WoTG and before   SoA.

RaraProjects committed:
Added Pining Nocturne (472) to spell_list.sql

RaraProjects committed:
Implemented pining_nocturne.lua

Frankie-hz commented on scripts/actions/spells/songs/pining_nocturne.lua:
```diff
@@ -0,0 +1,184 @@
+        end
+
+        -- Tier 1 Etudes -- Base power of 3.
+        if songEffect == xi.effect.ETUDE and songTier == 1 then
+            if     singingLvl >= 450 then power = power + 6
+            elseif singingLvl >= 397 then power = power + 5
+            elseif singingLvl >= 343 then power = power + 4
+            elseif singingLvl >= 289 then power = power + 3
+            elseif singingLvl >= 236 then power = power + 2
+            elseif singingLvl >= 182 then power = power + 1
+            end
+
+        -- Tier 2 Etudes -- Base power of 12.
+        elseif songEffect == xi.effect.ETUDE and songTier == 2 then
+            if     singingLvl >= 475 then power = power + 3
+            elseif singingLvl >= 446 then power = power + 2
+            elseif singingLvl >= 417 then power = power + 1
+            end
+
+        -- Check if we get the skill cap boost for other songs.
+        -- Ballad, Hymnus, Mazurka have constant base power.
+        -- Paeon can only get a total of +1 boost with capped skill.
+        else
```
This is going to be a very large if statement. I will let the others comment but you could use a table also.

```lua
local etudeTiers =
{
    [1] =
    {
        { level = 450, bonus = 6 },
        { etc } ...
    },
    [2] =
    {
        { level = 475, bonus = 3 },
        { etc } ...
    },
}

if
    songEffect == xi.effect.ETUDE and
    (songTier == 1 or songTier == 2)
then
    for _, tier in ipairs(etudeTiers[songTier]) do
        if singingLvl >= tier.level then
            power = power + tier.bonus
            break
        end
    end
else
    -- rest of code
end
```

RaraProjects replied:
I can definitely switch to a table if that is the preference of the team. I'm hoping you can answer a couple questions to help me build my intuition on this suggestion.

**Performance**
I talk to ChatGPT quite a bit. I asked if the table implementation or the IF implementation was faster. The difference is negligible but the IF statement is slightly faster (I didn't know this before just now). Is your concern coming from performance, presentation, or both? (I can change it to a table regardless of the reason. It just helps me better understand which style to use.)

**Long or Multiple Condition IF Statements**
I noticed you're using the multi line IF structure from the style guide for your entrance into the loop. If I was eyeing that up on my own I would think I could fit that one line. Should I be using the multi line IF statement if there are >2 conditions or is there kind of gray middle ground where either style is okay?

RaraProjects replied:
I've switched to the table format. I just need to test and I'll resubmit.

zach2good replied:
Performance: This calculation is done a single time on casting - don't worry about it 👍

WinterSolstice8 commented:
Sanity checks is failing from the commit name being too long. Amending the commit to use `Implement Pining Nocturne` for the message is fine

zach2good reviewed (commented):
Hello and welcome! 

It looks like you've put a lot of work and research into this, thanks, much appreciated!

I think GitHub is having a lot of trouble today, and our CI has been a little unreliable recently, but I'll give this a once-over - knowing that I don't know much about this spell and I've forgotten everything about being a Bard...

zach2good commented on scripts/actions/spells/songs/pining_nocturne.lua:
```diff
@@ -0,0 +1,184 @@
+    return songEffect
+end
+
+return spellObject
```
Newline at end of file please

zach2good commented on scripts/actions/spells/songs/pining_nocturne.lua:
```diff
@@ -0,0 +1,184 @@
+    -- Current state may be a precursor to general BRD song handling.
+
+    -- Soul voice affects different songs differently.
+    local soulVoiceBoostType = {
```
For clarity while reading these sections, you can extract these little tables of stuff (`strongAffinityAcc` too, etc.) to the parent scope

Xaver-DaRed replied:
Also, the style we use is
```lua
local tableName =
{
    Stuff
}

zach2good reviewed (commented):
Seems reasonable 👍

RaraProjects commented:
Thank you all for the great feedback. Here's a summary of what's been done with your feedback.

## Addressed Feedback
1. Comment too long.
2. ~~Skill cap table instead of IF statement.~~ (see comment 1 below)
3. New line at EOF.
4. Logistics tables moved to parent scope.
5. Table style updates.
6. TODO about Nocturne/Addle. I also switched all of my FUTURE to TODO.
7. Performed tests from the pull request text after changes. All passed with expected results.

## Additional Comments
1. @Xaver-DaRed I can get behind having separate handlers for the enhancing and enfeebling songs. Thank you for the extra background! I definitely don't want to mess up what you have going on right now. The next phase in my plan was to create a new issue where I would lay out my implementation in more detail. Given your comment I'll probably just focus on enfeebling songs and leave the enhancing songs alone entirely. Consequently, I've removed the skill handling (including the Etude skill cap) from my submission.
2. @Xaver-DaRed thanks for the background on magic.lua. This is great for helping me better understand the current state of the project. :)
3. @Xaver-DaRed I think the instrument skill issue is a rookie mistake on my part. I made a copy of LSB at the end of November and never updated it. I did my initial development on that copy. After consulting your Git guide and ChatGPT, I made a fork of LSB just recently. I copied my code from my original copy into my fork. You made the update to the skill on 12/09. So what you're seeing is old LSB code. Skill issue on my part. The old LSB shouldn't be a factor in the future now that I'm getting a handle on Git and how the universe is woven together.
4. Thanks for the warm welcome! :)

claywar commented on scripts/actions/spells/songs/pining_nocturne.lua:
```diff
@@ -0,0 +1,150 @@
+    power = power + (gearBoost * powerBoostMult)    -- Boost from Song+ and All_Song+ gear.
+
+    -- Add power from merits.
+    if meritEffect ~= 0 then
```
This is dead code, as you're initializing to 0, and nothing prior is modifying it

RaraProjects replied:
I'm leaving it in there as a placeholder for the enfeeble song follow up work. Eventually I'm planning to move this code into a general enfeeble song function that handles all enfeebling songs. Currently, it doesn't look like enfeeble songs benefit from merit points in retail, but having this hook available leaves the door open for any future private servers  based on LSB to add custom merit points to affect enfeeble songs.

WinterSolstice8 replied:
Nah, don't need to worry about that. That's what Lua modules are for

RaraProjects replied:
I'll remove this.

claywar commented on scripts/actions/spells/songs/pining_nocturne.lua:
```diff
@@ -0,0 +1,150 @@
+    end
+
+    -- Add power from job points.
+    if jobPointEffect ~= 0 then
```
Similar to above, dead code

RaraProjects replied:
Similar to the comment about merits. Except for the job points there are retail boosts for enfeebles (Lullaby duration and Requiem effect). Looking at this, I'm realizing there is a job point bonus for song accuracy that I'm not taking into account right now, but maybe I can add that in the follow up work.

RaraProjects replied:
I'm going to add a TODO for the job point song accuracy bonus (xi.jp.SONG_ACC_BONUS) and for the job point gifts of magic accuracy. I'll restrict the job point power boost to requiem since that is the only spell affected in retail. I'm also adding a x3 just like the requiem luas. I was missing a hook for job point effect on duration entirely so I added that and restricted it to just lullaby. Thanks for your feedback!

RaraProjects commented:
All comments up to this point should be taken care of and the changes have been pushed. Please let me know if there are any other concerns! :)

Xaver-DaRed commented on scripts/actions/spells/songs/pining_nocturne.lua:
```diff
@@ -0,0 +1,151 @@
+    local dotTick         = 0
+
+    -- Check to see if the player has any elemental affinity gear like an elemental staff.
+    local elementalAccBonus = caster:getMod(strongAffinityAcc[spell:getElement()]) * 10
```
Time for nitpicks. Lets go...
```lua
local actionElement     = spell:getElement()
local elementalAccBonus = caster:getMod(xi.combat.element.elementalMagicAcc[actionElement]) + caster:getMod(xi.combat.element.strongAffinityAcc[actionElement]) * 10
```
You have elemental macc from certain gear (not the staves), like straps.
Also, this makes it so you can delete the table you defined up there

RaraProjects replied:
It's the small things that add the polish. This is a great addition. I was considering using xi.combat.element.strongAfinityAcc, but was unsure how to utilize it.

Out of scope for this work, but is there there an opportunity to create a getMagicAccuracy lua function that could be used to get both the magic accuracy and affinity accuracy in one step? Or would that not be worth the effort?

Xaver-DaRed commented on scripts/actions/spells/songs/pining_nocturne.lua:
```diff
@@ -0,0 +1,151 @@
+    end
+
+    -- TODO: Power calculation can be modularized.
+    power = math.min(power, powerCap)               -- Cap power if necessary. (Applied before Merits, Job-Points and Status-Effects)
```
whenever you can, its better to do this:
`power = utils.clamp(power, 0, powerCap)`
Why? So there is 0 chance to go negative. Its also visually easier to know the allowed range we want the value to be.

Xaver-DaRed commented on scripts/actions/spells/songs/pining_nocturne.lua:
```diff
@@ -0,0 +1,151 @@
+
+    -- TODO: Power calculation can be modularized.
+    power = math.min(power, powerCap)               -- Cap power if necessary. (Applied before Merits, Job-Points and Status-Effects)
+    power = power + (gearBoost * powerBoostMult)    -- Boost from Song+ and All_Song+ gear.
```
This applies to other lines, not just this one:
`power = power + gearBoost * powerBoostMult`

Unnecesary parenthesis are unnecesary.

RaraProjects commented:
Additional feedback has been implemented and pushed. I made one additional change. I removed the "0" tier parameter from the addStatusEffect call. I didn't see a purpose for it the enfeebling song context.
